src/iam-token.js

```diff
@@ -0,0 +1,146 @@
+    } else if (typeof apiKeyRef == 'object') {
+      let secretName = objectPath.get(apiKeyRef, 'valueFrom.secretKeyRef.name');
+      let secretNamespace = objectPath.get(apiKeyRef, 'valueFrom.secretKeyRef.namespace', namespace);
+      let secretKey = objectPath.get(apiKeyRef, 'valueFrom.secretKeyRef.key');
+      apiKey = await this._getSecretData(secretName, secretKey, secretNamespace, kubeResourceMeta, namespace);
+    }
+    if (!apiKey) {
+      return Promise.reject('Failed to find valid apikey to authenticate against iam');
+    }
+
+    let res = await this._requestToken(iam, apiKey);
+    return res.access_token;
+  }
+
+
+  async _requestToken(iam, apiKey) {
+    let token;
+
+    const apiKeyHash = hash(apiKey, { algorithm: 'shake256' });
+    // Check if the cache has the token. If it does have it, copy it into token
+    // for further processing
+    if (objectPath.has(this.s3TokenCache, [apiKeyHash])) {
+      // fetch cached token
+      token = objectPath.get(this.s3TokenCache, [apiKeyHash]);
+    }
+
+    // if we are able to fetch the token from the cache and it is not a Promise,
+    // then we need to check the expiry. If it is not expiring in the next 120 s
+    // return the token
+    if (!(token instanceof Promise)) {
+      if (token !== undefined) {
+        const expires = objectPath.get(token, 'expiration', 0); // expiration: time since epoch in seconds
+        // re-use cached token as long as we are more than 2 minutes away from expiring.
+        if (Date.now() < (expires - 120) * 1000) {
+          return token;
+        }
+      } else {
+        // if we get to this point the token is either undefined or it is stale data.
+        // either ways we need to connect to iam auth and get a new token
+        // create an async function that will wait for _getToken to return a token.
+        // since we are not using await, the control will return immediately and the
+        // return value of the tokenFunction is going to be a promise.
+        // we are going to cache this promise into the s3TokenCache and
+        // wait for it.
+        let tokenFunction = async (iam, apiKeyHash, apiKey) => {
+          let ret = await this._getToken(iam, apiKeyHash, apiKey);
+          return ret;
+        };
+        token = tokenFunction(iam, apiKeyHash, apiKey);
+        objectPath.set(this.s3TokenCache, [apiKeyHash], token);
+      }
+    }
+
+    // at this point token is either data, or Promise (either created in the else case above
+    // or already created by a different event)
+    if (token instanceof Promise) {
+      // if the token is a Promise, wait for it to be completed. Once completed, get
+      // the data from the cache and return it.
+      try {
+        await token;
+        return objectPath.get(this.s3TokenCache, [apiKeyHash]);
+      } catch(error) {
+        Promise.reject('failed to get the new token:', error);
```

Comment:
```suggestion
        return Promise.reject('failed to get the new token:', error);
```
Need to return the Promise.reject(), or just throw an error.  _just_ calling `Promise.reject(...)` won't have the desired effect.

Reply:
Fixed. Thanks!